wire-runtime/src/main/java/com/squareup/wire/Message.java

```diff
@@ -301,4 +261,244 @@ protected IllegalStateException missingRequiredFields(Object... args) {
      */
     public abstract T build();
   }
+
+  protected static <M extends Message<M>> M message(ProtoReader reader, TypeAdapter<M> adapter)
+      throws IOException {
+    long cursor = reader.beginLengthDelimited();
+    M message = adapter.read(reader);
+    reader.endLengthDelimited(cursor);
+    return message;
+  }
+
+  protected static <M extends Message<M>> List<M> repeatedMessage(List<M> existing,
+      ProtoReader reader, TypeAdapter<M> adapter) throws IOException {
+    long cursor = reader.beginLengthDelimited();
+    existing = repeated(existing, adapter.read(reader));
+    reader.endLengthDelimited(cursor);
+    return existing;
+  }
+
+  protected static <M extends Enum<M> & WireEnum> M enumOrUnknown(int tag, ProtoReader reader,
+      TypeAdapter.EnumAdapter<M> adapter, Builder builder) throws IOException {
+    int value = reader.value(TypeAdapter.UINT32);
+    M constant = adapter.fromValue(value);
+    if (constant != null) {
+      return constant;
+    }
+    builder.addVarint(tag, value);
+    return null;
+  }
+
+  protected static <M extends Enum<M> & WireEnum> List<M> repeatedEnumOrUnknown(int tag,
```

Comment:
Does this method handle enums and unknowns both together because they're the same? Or just because the code to support them is the same? Maybe would be neater to have separate methods like `repeatedEnum()`, `repeatedUnknown()` and for those to delegate to this, which would be private?